### ISSUE TITLE
library or app-with-plugins

### ISSUE THREAD

**Issue by cgwalters** [purpleidea/oh-my-vagrant#80]
The current design of OMV seems to be heading towards the "app with plugins" model.  The UX here is that a user `git clone`s OMV, then can further choose plugins like `examples/kubernetes-ansible.yaml`.

Without having played with this very much (so take with a grain of salt) I'm feeling like it'd be better if OMV was a shared library, specifically if it was designed to be used as a git submodule.

The thing is - what I'd like to have for Project Atomic is a git repository that when `git clone`d from master, is known to work starting up an Atomic cluster. With the current model as OMV expands with more functionality, it seems quite possible for that this will get harder to maintain.  There are so many factors in the mix here - base box type (CentOS/Fedora) base box upgrade mechanism (traditional, atomic), target functionality (k8s, puppet, other?)

Thoughts?

**Comment by purpleidea** [comment]
@cgwalters Thanks for your comments...

Let me try to explain the way it's used, and then offer some suggestions on what could be done to improve this. In the meantime, I urge you to try the tool a bit more, and/or checkout some of the screencasts I've made...

The feature you want of "is a git repository that when git cloned from master, is known to work starting up an Atomic cluster" will work and should continue to always work, as soon as the issues in #78 are resolved. The files in examples/*.yaml aren't plugin, but rather definition files. So the two steps for reproducing are: 1) clone omv, 2) copy in a particular omv.yaml file. You have this feature today, minus a few Atomic base image bugs.


Having said all that, you talk about turning omv into a "shared library" of sorts. Here are some ways that could work:

1) libify this into an actual ruby library.
This would take code out of the Vagrantfile and be put somewhere else. This might actually be useful, but it might be harder to distribute.

2) turn omv into a vagrant plugin
This  would actually be pretty useful, however it's a lot of work, with relatively little gain.

3) turn some of the internal codified omv features into vagrant-plugins
This is something that would be very reasonable to do. Taking the kubernetes support that I wrote, and turning that into a native provisioner would be a sane thing to do. I actually mention this in: https://github.com/purpleidea/oh-my-vagrant/blob/master/vagrant/Vagrantfile#L956

So I recommend doing (3) if someone else wanted to do it. It is done the way it currently is, because one metric I used was "most functionality for least development work"... (2) would be great, but it would take 5 times as much work to get it feature complete with what already works. It would then also be harder for others to modify, which is another downside. (1) isn't very useful I don't think.

HTH,
James
